### PR TITLE
Feature/random number endpoint

### LINKED ISSUE #9: Random Number endpoint
- max
- min

### PR DESCRIPTION
**Pull request by ntiany**

### CODE REVIEW

File: RandomData.Api/NumberGenerators/Dtos/NumberParameters.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using Microsoft.AspNetCore.Mvc;
+
+namespace RandomData.Api.NumberGenerators.Dto
```
Comment (KanarekLife) [inline RandomData.Api/NumberGenerators/Dtos/NumberParameters.cs]: Name of folder is 'Dto's so namespace should be "RandomData.Api.NumberGenerators.Dtos"

File: RandomData.Api/NumberGenerators/Dtos/NumberParameters.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public class NumberParameters
+    {
+        private const string DefaultAllowedDigits = "1234567890";
```
Comment (KanarekLife) [inline RandomData.Api/NumberGenerators/Dtos/NumberParameters.cs:7]: If I were you I would also add options such as:
- min value
- max value
- base (10, 2, 4, 8 or more)

File: RandomData.Api/NumberGenerators/Generators/INumberGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
```
Comment (KanarekLife) [inline RandomData.Api/NumberGenerators/Generators/INumberGenerator.cs]: Unnecessary import of these System libraries.

File: RandomData.Api/NumberGenerators/Generators/NumberGenerator.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
```
Comment (KanarekLife) [inline RandomData.Api/NumberGenerators/Generators/NumberGenerator.cs]: Unnecessary import

File: RandomData.Api/NumberGenerators/Generators/NumberGenerator.cs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+using System.Linq;
+using System.Threading.Tasks;
```
Comment (KanarekLife) [inline RandomData.Api/NumberGenerators/Generators/NumberGenerator.cs]: Unnecessary import

File: RandomData.Api/NumberGenerators/Validators/NumberParametersValidator.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using System.Linq;
+using FluentValidation;
+using RandomData.Api.NumberGenerators.Dto;
+using RandomData.Api.NumberGenerators.Validators;
+
+namespace RandomData.Api.NumberGenerators.Validator
```
Comment (KanarekLife) [inline RandomData.Api/NumberGenerators/Validators/NumberParametersValidator.cs]: Should be: 'RandomData.Api.NumberGenerators.Validators'

File: RandomData.Api/NumberGenerators/Validators/NumberParametersValidator.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+{
+    public class NumberParametersValidator : AbstractValidator<NumberParameters>, INumberParametersValidator
+    {
+        public NumberParametersValidator()
+        {
+            RuleFor(x => x.MinLength).GreaterThanOrEqualTo(1);
+            RuleFor(x => x.MaxLength).LessThanOrEqualTo(int.MaxValue);
+            RuleFor(x => x.MaxLength).GreaterThanOrEqualTo(x => x.MinLength);
```
Comment (KanarekLife) [inline RandomData.Api/NumberGenerators/Validators/NumberParametersValidator.cs]: I would personally write this way:
```cs
RuleFor(x => x.AllowedDigits)
                .MinimumLength(1)
                .Must(value => value.All(char.IsDigit));
```

File: RandomData.Api/Startup.cs [hunk 8]
```diff
@@ -20,6 +21,7 @@ public Startup(IConfiguration configuration)
 		public void ConfigureServices(IServiceCollection services)
 		{
 			services.AddGuidGenerator();
+            services.AddNumberGenerator();
```
Comment (KanarekLife) [inline RandomData.Api/Startup.cs]: These should be in one line.

File: RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        {
+            //arrange
+            var generator = new NumberGenerator();
+            var parameters = new NumberParameters();
+
+            //act
+            var result = generator.GetRandom(parameters);
+
+            //assert
+            Assert.True(result >= 0);
```
Comment (KanarekLife) [inline RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs:20]: We decided to use FluentValidation so it should be:
```cs
result.Should().BeGreaterOrEqualTo(0);
```
instead of 
```cs
Assert.True(result >= 0);
```

File: RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs [hunk 10]
```diff
@@ -0,0 +1,95 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using RandomData.Api.NumberGenerators.Dto;
+using RandomData.Api.NumberGenerators.Generators;
+using Xunit;
+
+namespace RandomData.Api.Tests.NumberGenerators
+{
+    public class NumberGeneratorTests
```
Comment (KanarekLife) [inline RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs:7]: I'm afraid that in an event where something breaks we might not see the problem right away. The tests are not repeatable because they are using randomness, so for example if max value cap does not work, the test can still pass due to randomness.

File: RandomData.Api.Tests/NumberGenerators/NumberParametersValidatorTests.cs [hunk 11]
```diff
@@ -0,0 +1,78 @@
+﻿using RandomData.Api.NumberGenerators.Dto;
+using RandomData.Api.NumberGenerators.Validator;
+using Xunit;
+
+namespace RandomData.Api.Tests.NumberGenerators
+{
+    public class NumberParametersValidatorTests
```
Comment (KanarekLife) [inline RandomData.Api.Tests/NumberGenerators/NumberParametersValidatorTests.cs:7]: I don't think that packages should be tested by us. I think they should be tested by package maintainer.

File: RandomData.Api/NumberGenerators/Validators/INumberParametersValidator.cs [hunk 12]
```diff
@@ -0,0 +1,13 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using FluentValidation;
+using RandomData.Api.NumberGenerators.Dto;
+
+namespace RandomData.Api.NumberGenerators.Validators
+{
+    public interface INumberParametersValidator : IValidator<NumberParameters>
```
Comment (KanarekLife) [inline RandomData.Api/NumberGenerators/Validators/INumberParametersValidator.cs:6]: I personally do not see the point but I might also not understand it  ¯\\_(ツ)_/¯

**Review by KanarekLife — CHANGES_REQUESTED** [review]
I would change a few things but wait for @krzysztof-gwozdz opinion ;)

File: RandomData.Api/NumberGenerators/Validators/INumberParametersValidator.cs (hunk 12 above)
Comment (ntiany) [reply RandomData.Api/NumberGenerators/Validators/INumberParametersValidator.cs:6]: I saw it separated in Guid implementations and I wanted to keep it consistent

File: RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs (hunk 10 above)
Comment (ntiany) [reply RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs:7]: we might consider repeating generating numbers several times in a loop (f.e. 100 times) to minimize randomness in a given test

File: RandomData.Api.Tests/NumberGenerators/NumberParametersValidatorTests.cs (hunk 11 above)
Comment (ntiany) [reply RandomData.Api.Tests/NumberGenerators/NumberParametersValidatorTests.cs:7]: I don't test packages, I test my rules - if they are defined correctly

File: RandomData.Api/NumberGenerators/Dtos/NumberParameters.cs (hunk 2 above)
Comment (ntiany) [reply RandomData.Api/NumberGenerators/Dtos/NumberParameters.cs:7]: I will consider adding it in a separate route
min and max value won't work well with length and allowed digits parameters, it will make validation and logic very complicated

Comment (KanarekLife) [reply RandomData.Api/NumberGenerators/Dtos/NumberParameters.cs:7]: Well currently it works almost identical to string endpoint with just numbers specified so that's why I would focus more on stuff that cannot be achieved elsewhere

File: RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs (hunk 10 above)
Comment (KanarekLife) [reply RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs:7]: At first I had simmilar idea in my endpoint but it's not a way to go. You should create edge points that you suspect may break something and have something like mine FakeRandom to explicitly test these edge cases. I don't think you should ever use random methods in tests because they should be repetitive.

File: RandomData.Api.Tests/NumberGenerators/NumberParametersValidatorTests.cs (hunk 11 above)
Comment (KanarekLife) [reply RandomData.Api.Tests/NumberGenerators/NumberParametersValidatorTests.cs:7]: What I mean by testing packages is that you should assume that whenever you do something in package it'll work correctly. So if you create a rule that X must be lower than 10 you are sure that it will be correctly validated, because the tests to test whether it works are supposed to be in the package source files and thus should not be your concern.

I also might be wrong, so @krzysztof-gwozdz need to verify this ;)

Comment (ntiany) [reply RandomData.Api.Tests/NumberGenerators/NumberParametersValidatorTests.cs:7]: I still test my rules, if nothing is missing in sets defined by me. I understand what you say, I just don't agree :)

File: RandomData.Api/NumberGenerators/Dtos/NumberParameters.cs (hunk 2 above)
Comment (ntiany) [reply RandomData.Api/NumberGenerators/Dtos/NumberParameters.cs:7]: I kept to what we determined during our meeting - max and min length for a number endpoint

File: RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs (hunk 10 above)
Comment (ntiany) [reply RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs:7]: I don't think it's necessary to introduce fake random, probability of receiving 100 times correct answer by a coincidence is so low you can ignore it

Comment (ntiany) [reply RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs:7]: and it's kinda bad to modify implementation just for testing purposes

Comment (KanarekLife) [reply RandomData.Api.Tests/NumberGenerators/NumberGeneratorTests.cs:7]: I think that black-box techniques should only be used to as an addition not as  an only tests of an component. 

I think it @krzysztof-gwozdz might give his thought

**Review by lucaslov — APPROVED** [review]

**Review by mejtika — APPROVED** [review]
